### PR TITLE
ci: add `lint-skill-entry` structural validator

### PR DESCRIPTION
**Pull request by MajorLift**
## Summary

Acceptance criteria for skill contributions, in both halves — a CI validator for the deterministic properties, and the two questions in `CONTRIBUTING.md` that no check can answer. From the [ADR 0057 review](https://github.com/MetaMask/decisions/pull/162), where acceptance criteria were raised as an open item.

## The validator

`yarn lint:skills` checks directory layout, `skill.md` path and name pattern, frontmatter keys and maturity values, `mms-` prefix misuse, repo-overlay names, and description length. `tools/skill-schema.mjs` is the single source of truth, imported by the linter so the documented and enforced schemas cannot drift.

**Description budget is 1,536** — a repo budget, not an operator limit.

An earlier draft used 1,024, justified as the per-operator minimum citing OpenCode. That does not hold up: `tools/install` emits the description verbatim with no truncation anywhere in the tooling, and six skills over 1,024 — up to 1,344 — install and load in Claude Code today. It also never bound anything, since the 46 skills on `main` have a median description of 45 characters and a maximum of 928.

Enforcing it had a real cost. A description is the discovery surface and the only part of a skill carrying its own trigger cues, so trimming one to fit makes the skill less likely to be selected when it is relevant. The budget still exists — always-on context should be bounded — but at a number that reflects a deliberate choice rather than an unverified external claim, and the comment now asks for an operator and version before anyone tightens it.

## What running it against `main` turned up

Three errors, two of which were defects in the repo rather than in the check:

- **`workflows/` never shipped.** 14 files across two `web3-tools` skills, referenced **22 times** from their bodies, absent from the installer's bundle list — so every installed copy carried 22 dangling links and nothing reported it. Added to `copy_bundle_dirs`; all 22 now resolve, 42 files delivered.

  This is the third instance of one root cause, after `pr-validate`'s `hooks/` and domain `knowledge/`: a directory that exists in source and is not in the copy list.

- **`metadata` was rejected** as unknown frontmatter, though README documents it as preserved through install. Now in the optional set.

- **`performance` description was 1,078 chars.** Under the 1,536 budget, so it stays as authored — an earlier trim to 928 was reverted along with the ceiling that motivated it.

Validator now reports **0 errors across 46 skills**.

## So the list stops being written twice

Two tests read `tools/install` directly:

- `BUNDLE_DIRS` must match the directories it copies.
- Every key it reads via `frontmatter_value` must be declared in the schema.

The comment asking a human to keep Bash and JS in sync is now checked. Knowledge files also get their own declared keys, since they take `domain` rather than the installer-behaviour keys skills use.

## And the half a linter cannot check

`CONTRIBUTING.md` gains two questions:

**Is it a skill, or does it belong in an enforcement layer?** Skills shape generation; linters, hooks, and fitness functions enforce. A skill that *substitutes* for enforcement is unsafe — an agent can ignore any context it is given. One that *restates* a deterministic check is wasteful. One that *teaches the upstream pattern* so enforcement rarely fires is the right shape, and should name the layer it pairs with.

**Does it earn its context budget?** Frontmatter for every installed skill loads at agent startup, so a skill that is never selected still costs its `description` on every run.

## Keeping the number honest

Two tests, because the failure mode here was documentation drifting from enforcement in both directions:

- The boundary case derives its length from `DESCRIPTION_MAX` rather than hardcoding one, so it keeps testing the boundary when the budget moves. The previous hardcoded 1,100 silently stopped exercising anything the moment the ceiling rose.
- A new check fails if `README.md`, `CONTRIBUTING.md`, or `.github/SKILL_TEMPLATE.md` state a figure the schema does not enforce. All three had said 1,536 while the schema said 1,024; reconciling them by editing the docs down was the wrong direction, and this makes that disagreement fail rather than get silently resolved.

## Test plan

- [x] `node .github/scripts/lint-skill-entry.mjs` — 0 errors, 46 skills
- [x] `test/lint-skill-entry.test.mjs` — 10 pass, including both drift checks
- [x] `test/cli.test.mjs` 11 pass, `test/unit.test.mjs` 30 pass
- [x] Real install of `web3-tools`: 22/22 `workflows/` references resolve, 42 files delivered
- [x] 84 remaining warnings are all missing recommended sections — advisory, non-blocking

## Notes

- Supersedes #92, folded here: the check handles what it can, the prose says where to spend review attention instead.
- 84 warnings suggest either the recommended sections are wrong for this corpus or most skills predate them. Worth a follow-up decision; not blocking, since warnings do not fail CI.

### CODE REVIEW

File: .github/workflows/lint-skill-entry.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+      - name: Get changed skill files
+        id: changed
+        uses: tj-actions/changed-files@v45
+        with:
+          files: domains/**
+
+      - name: Lint changed skills
+        if: steps.changed.outputs.any_changed == 'true'
+        run: node .github/scripts/lint-skill-entry.mjs ${{ steps.changed.outputs.all_changed_files }}
```
Comment (NicolasMassart) [inline .github/workflows/lint-skill-entry.yml]: **issue (blocking, security):** Changed filenames are interpolated directly into a shell command.

`${{ steps.changed.outputs.all_changed_files }}` is space-joined and dropped into the shell. Filenames come from PR contents, so a path with spaces splits incorrectly and a path with shell metacharacters can alter the command. `contents: read` reduces blast radius but does not make this safe. Resolved by the suggested step on line 16 — files are read via `git diff -z`, passed through an env var, and parsed by `mapfile`/argv rather than interpolated.

```yaml
      - name: Lint changed skills
        if: steps.changed.outputs.files != ''
        env:
          CHANGED_FILES: ${{ steps.changed.outputs.files }}
        run: |
          mapfile -t files <<< "$CHANGED_FILES"
          node .github/scripts/lint-skill-entry.mjs "${files[@]}"
```

File: .github/scripts/lint-skill-entry.mjs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  const resolved = paths.map((file) => path.resolve(ROOT, file));
+  return skills.filter((skill) =>
+    resolved.some((file) => file === skill.path || file.startsWith(`${skill.path}${path.sep}`)),
+  );
+}
+
+function main() {
+  const paths = process.argv.slice(2).filter((arg) => !arg.startsWith('-'));
+  const all = collectSkills([ROOT]);
+  const skills = paths.length > 0 ? skillsForPaths(all, paths) : all;
```
Comment (NicolasMassart) [inline .github/scripts/lint-skill-entry.mjs]: **issue (blocking):** Malformed directory shapes silently pass — the gate doesn't actually enforce `domains/<domain>/skills/<name>/skill.md`.

The PR description claims to enforce that layout, but `main()` calls `collectSkills([ROOT])` first and *then* filters by changed paths. `collectSkills` only returns directories that already match the expected layout *and* contain a readable `skill.md`; anything malformed is invisible to `skillsForPaths` and the run prints `0 skill(s) checked, 0 error(s)`. All of these slip through:

```text
domains/foo/bar/skill.md                       # missing skills/<name>/ segment
domains/foo/skills/bar/SKILL.md                # wrong filename casing
domains/foo/skills/bar/README.md               # skill.md missing entirely
domains/foo/skills/bar/references/doc.md       # changed file under skill-shaped dir, no skill.md
```

The last one is the worst: a changed file lives under a skill-shaped directory but the linter has no skill object to validate, so a `skill.md` deletion or rename inside a single PR ships unchecked.

Adding a path-level validation phase before the `collectSkills` filter closes this. For every changed `domains/**` path, derive the expected skill root and fail if (a) the path is not under `domains/<domain>/skills/<name>/...`, or (b) the root has no readable `skill.md`. Sketch:

```javascript
const SKILL_PATH = /^domains\/([^/]+)\/skills\/([^/]+)\/(?:[^/]+\/)*[^/]+$/u;

function validatePathShape(file) {
  if (!file.startsWith('domains/')) return null;
  const m = SKILL_PATH.exec(file);
  if (!m) return `path "${file}" is not under domains/<domain>/skills/<name>/`;
  const [, domain, name] = m;
  const root = path.join(ROOT, 'domains', domain, 'skills', name);
  try {
    statSync(path.join(root, 'skill.md'));
  } catch {
    return `skill root "domains/${domain}/skills/${name}/" has no readable skill.md`;
  }
  return null;
}

function main() {
  const paths = process.argv.slice(2).filter((arg) => !arg.startsWith('-'));

  let errorCount = 0;
  for (const file of paths) {
    const err = validatePathShape(file);
    if (err) {
      console.log(`\nerror: ${err}`);
      errorCount += 1;
    }
  }

  const all = collectSkills([ROOT], '*');
  const skills = paths.length > 0 ? skillsForPaths(all, paths) : all;
  // ... existing per-skill loop, accumulating into errorCount ...
  process.exitCode = errorCount > 0 ? 1 : 0;
}
```

File: test/lint-skill-entry.test.mjs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  return dir;
+}
+
+function lint(root) {
+  const result = spawnSync(process.execPath, [LINTER], {
+    env: { ...process.env, SKILLS_LINT_ROOT: root },
+    encoding: 'utf8',
+  });
+  return { code: result.status, output: `${result.stdout}${result.stderr}` };
+}
```
Comment (NicolasMassart) [inline test/lint-skill-entry.test.mjs]: **issue (blocking, test):** Tests cover full-catalogue mode only; the CI-critical changed-files mode has zero coverage.

`lint(root)` invokes the linter with no path arguments, so every test exercises the `paths.length === 0` branch (full audit). The workflow runs the *other* branch — `node .github/scripts/lint-skill-entry.mjs <changed files...>` — and that path has no tests. Combined with the structural-check gap, that is how malformed paths reach `main`. Add cases that mirror the CI invocation:

- valid changed `skill.md` is linted
- a changed reference file (`references/foo.md`) maps back to its skill root
- malformed `domains/**` path fails (`domains/foo/bar/skill.md`)
- a deleted/missing `skill.md` is reported, not silently skipped
- filenames containing spaces survive argv handling
- warnings-only input exits 0
- a description of exactly `DESCRIPTION_MAX` passes; `+1` fails

```javascript
function lint(root, paths = []) {
  const result = spawnSync(process.execPath, [LINTER, ...paths], {
    env: { ...process.env, SKILLS_LINT_ROOT: root },
    encoding: 'utf8',
  });
  return { code: result.status, output: `${result.stdout}${result.stderr}` };
}

test('changed skill.md is linted (CI mode)', () => {
  const root = makeRoot();
  writeSkill(root, 'testing', 'unit-testing', 'name: unit-testing\ndescription: x');
  const file = path.join(root, 'domains/testing/skills/unit-testing/skill.md');
  assert.equal(lint(root, [file]).code, 0);
});

test('malformed changed path fails before collectSkills', () => {
  const root = makeRoot();
  const stray = path.join(root, 'domains/foo/bar/skill.md');
  mkdirSync(path.dirname(stray), { recursive: true });
  writeFileSync(stray, '---\nname: bar\ndescription: x\n---\n');
  const { code, output } = lint(root, [stray]);
  assert.equal(code, 1);
  assert.match(output, /not under domains\/<domain>\/skills/u);
});

test('description at the ceiling passes; +1 fails', () => {
  const ok = makeRoot();
  writeSkill(ok, 'testing', 'unit-testing', `name: unit-testing\ndescription: ${'x'.repeat(1024)}`);
  assert.equal(lint(ok).code, 0);
  const bad = makeRoot();
  writeSkill(bad, 'testing', 'unit-testing', `name: unit-testing\ndescription: ${'x'.repeat(1025)}`);
  assert.equal(lint(bad).code, 1);
});
```

File: .github/workflows/lint-skill-entry.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 24.x
+
+      - name: Get changed skill files
+        id: changed
+        uses: tj-actions/changed-files@v45
```
Comment (NicolasMassart) [inline .github/workflows/lint-skill-entry.yml]: **suggestion (security):** Pin `tj-actions/changed-files` to a commit SHA, or replace it.

This is the only third-party action under `.github/workflows/`. The repo's other workflows use MetaMask-owned actions (`MetaMask/action-checkout-and-setup@v3`) and the actionlint download in `main.yml` is pinned to a specific SHA — same posture is expected here, especially given v45's compromise history (March 2025). Replacing the action with `git diff -z` (see line 16) removes this concern entirely; if the action stays, pin to a SHA and add a Dependabot entry.

File: .github/workflows/lint-skill-entry.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+name: Lint skill entries
+
+on:
+  pull_request:
+    paths:
+      - 'domains/**'
+      - 'tools/**'
```
Comment (NicolasMassart) [inline .github/workflows/lint-skill-entry.yml:7]: **suggestion:** `tools/**` is in the trigger paths but never reaches the lint step.

The workflow triggers on `tools/**`, but the changed-files filter on line 29 lists only `domains/**`. So a `tools/skill-schema.mjs` change spins the workflow up, computes zero changed skill files, skips the lint step, and reports green — the catalogue is never re-validated against the tightened schema. Either drop `tools/**` from the trigger or branch on it: when `tools/skill-schema.mjs` is in the changed set, run a full audit (`node .github/scripts/lint-skill-entry.mjs` with no args) so a schema change re-checks every existing skill. The full-audit mode is what catches drift like the `workflows/` directories already on `main`.

File: tools/skill-schema.mjs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+// does not ship it and the reference would dangle post-install.
+export const ALLOWED_SIBLING_DIRS = [...BUNDLE_DIRS, 'repos'];
+
+export const KNOWN_REPOS = ['metamask-extension', 'metamask-mobile', 'core'];
+
+// The description is always loaded into the operator's discovery surface, so it
+// is the per-skill always-on cost. The ceiling is the per-operator minimum
+// (OpenCode caps description at 1024), so a description that passes here is
+// accepted by every target.
+export const DESCRIPTION_MAX = 1024;
```
Comment (NicolasMassart) [inline tools/skill-schema.mjs]: **issue (non-blocking, docs):** `README.md` documents `≤1,536 chars` but this enforces `1024`.

`DESCRIPTION_MAX = 1024` is the OpenCode minimum and the right ceiling, but the authoring section in `README.md` (line 369) still says descriptions may be `≤1,536 chars`. PR #46 has authors return to the README as canonical, so a contributor following the docs can land a skill that this validator rejects. Either update `README.md` here (one-line change to align with `DESCRIPTION_MAX`) or land the docs reconciliation before enabling the gate.

File: .github/scripts/lint-skill-entry.mjs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  if (!raw.description) {
+    errors.push('missing required `description` in frontmatter');
+  } else if (raw.description.length > DESCRIPTION_MAX) {
+    errors.push(`\`description\` is ${raw.description.length} chars, over the ${DESCRIPTION_MAX}-char operator ceiling`);
+  }
+
+  if (raw.maturity && !MATURITY_VALUES.includes(raw.maturity)) {
+    errors.push(`\`maturity\` "${raw.maturity}" must be one of: ${MATURITY_VALUES.join(', ')}`);
+  }
```
Comment (NicolasMassart) [inline .github/scripts/lint-skill-entry.mjs:71]: **suggestion:** Validate `scope` and `mandatory` — typos silently change semantics.

`OPTIONAL_FRONTMATTER` includes `scope` and `mandatory` and the installer treats `scope: user` specially (otherwise falls back to project scope). A typo like `scope: users` or `mandatory: ture` is currently a no-op — `KNOWN_FRONTMATTER` accepts the key and no enum validation runs, so the skill installs in a way the author didn't intend. Same shape as the existing `maturity` check; warnings are sufficient if you want to keep the blocking surface small.

```javascript
  if (raw.maturity && !MATURITY_VALUES.includes(raw.maturity)) {
    errors.push(`\`maturity\` "${raw.maturity}" must be one of: ${MATURITY_VALUES.join(', ')}`);
  }

  if (raw.scope !== undefined && !SCOPE_VALUES.includes(raw.scope)) {
    warnings.push(`\`scope\` "${raw.scope}" is unrecognised (known: ${SCOPE_VALUES.join(', ')}); installer falls back to project scope`);
  }

  if (raw.mandatory !== undefined && !TRUTHY.has(String(raw.mandatory).toLowerCase()) && !FALSY.has(String(raw.mandatory).toLowerCase())) {
    warnings.push(`\`mandatory\` "${raw.mandatory}" is not a recognised boolean (true|false|yes|no|1|0)`);
  }
```

File: .github/scripts/lint-skill-entry.mjs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+function skillsForPaths(skills, paths) {
+  const resolved = paths.map((file) => path.resolve(ROOT, file));
+  return skills.filter((skill) =>
+    resolved.some((file) => file === skill.path || file.startsWith(`${skill.path}${path.sep}`)),
+  );
+}
+
+function main() {
+  const paths = process.argv.slice(2).filter((arg) => !arg.startsWith('-'));
+  const all = collectSkills([ROOT]);
```
Comment (NicolasMassart) [inline .github/scripts/lint-skill-entry.mjs]: **suggestion (non-blocking):** Pass an explicit `repo` arg to `collectSkills`.

`collectSkills(sources, repo)` works incidentally with `repo` undefined because the linter never reads `repoApplicable`. A future refactor that makes `repo` required will break this silently. Pass an explicit sentinel and add a one-line note, or expose a no-repo entry point in the installer.

```javascript
  // Linter does not read repoApplicable; the wildcard documents intent and
  // survives a future signature where repo becomes required.
  const all = collectSkills([ROOT], '*');
  const skills = paths.length > 0 ? skillsForPaths(all, paths) : all;
```

File: .github/scripts/lint-skill-entry.mjs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    if (!KNOWN_FRONTMATTER.includes(key) && key !== 'alwaysApply') {
+      warnings.push(`unknown frontmatter key "${key}"; operators silently ignore unrecognised keys (typo?)`);
+    }
+  }
+
+  for (const section of RECOMMENDED_SECTIONS) {
+    if (!new RegExp(`^#{1,4}\\s+${section}\\b`, 'imu').test(skill.body)) {
+      warnings.push(`missing recommended section "## ${section}"`);
+    }
+  }
```
Comment (NicolasMassart) [inline .github/scripts/lint-skill-entry.mjs:122]: **polish:** Recommended-section regex matches `## When To Use Cases` as `When To Use`.

`\b` after a multi-word section name only requires the next character not to be a word boundary, so `## When To Use Cases` satisfies `When To Use`. If the intent is exact, anchor to end-of-line; if prefix is fine, leave a comment explaining the choice.

```javascript
  for (const section of RECOMMENDED_SECTIONS) {
    if (!new RegExp(`^#{1,4}\\s+${section}\\s*$`, 'imu').test(skill.body)) {
      warnings.push(`missing recommended section "## ${section}"`);
    }
  }
```

File: package.json [hunk 10]
```diff
@@ -21,7 +21,8 @@
     "test": "node --test test/*.test.mjs",
     "pack:dry-run": "yarn pack --dry-run",
     "lint": "yarn lint:changelog",
-    "lint:changelog": "auto-changelog validate --formatter oxfmt"
+    "lint:changelog": "auto-changelog validate --formatter oxfmt",
+    "lint:skills": "node .github/scripts/lint-skill-entry.mjs"
```
Comment (NicolasMassart) [inline package.json]: **suggestion (non-blocking):** `lint:skills` will fail on `main` post-merge — name signals scope.

The PR notes that the full audit currently surfaces drift (the `workflows/` directories in `smart-accounts-kit` / `x402-payments`, missing `maturity`, etc.) and is not blocking because the CI gate is changed-files only. That means `yarn lint:skills` will fail on `main` after merge until the retrofit lands. Either rename to `audit:skills` (matches the catalogue audit semantics) or split into `lint:skills` (changed-files mode, used by CI) and `audit:skills` (full-catalogue mode).

```json
    "audit:skills": "node .github/scripts/lint-skill-entry.mjs",
    "lint:skills": "node .github/scripts/lint-skill-entry.mjs --changed"
```

File: test/lint-skill-entry.test.mjs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  });
+
+  test('a knowledge/ sibling directory fails (the conversion guarantee)', () => {
+    const root = makeRoot();
+    const dir = writeSkill(root, 'perps', 'fix-bug', 'name: fix-bug\ndescription: x');
+    mkdirSync(path.join(dir, 'knowledge'));
+    const { code, output } = lint(root);
+    assert.equal(code, 1);
+    assert.match(output, /knowledge/u);
+  });
```
Comment (NicolasMassart) [inline test/lint-skill-entry.test.mjs:106]: **suggestion (test):** Add a `workflows/` sibling-directory fixture.

The current sibling-directory test uses `knowledge/`. The actual drift this PR's audit surfaces on `main` is the `workflows/` directory in `smart-accounts-kit` and `x402-payments` — that's the regression the gate is meant to prevent. A test against the live failure mode is worth more than the hypothetical one. Either replace or add alongside.

```javascript
  test('a workflows/ sibling directory fails (the live drift case)', () => {
    const root = makeRoot();
    const dir = writeSkill(root, 'perps', 'fix-bug', 'name: fix-bug\ndescription: x');
    mkdirSync(path.join(dir, 'workflows'));
    const { code, output } = lint(root);
    assert.equal(code, 1);
    assert.match(output, /workflows/u);
  });
```

File: tools/skill-schema.mjs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+// in tools/install (Bash) mirrors BUNDLE_DIRS; keep the two in sync.
+
+export const REQUIRED_FRONTMATTER = ['name', 'description'];
+export const OPTIONAL_FRONTMATTER = ['maturity', 'mandatory', 'scope'];
+export const KNOWN_FRONTMATTER = [...REQUIRED_FRONTMATTER, ...OPTIONAL_FRONTMATTER];
+
+export const MATURITY_VALUES = ['experimental', 'stable', 'deprecated'];
+
+// Directories the installer copies alongside skill.md (see tools/install).
+export const BUNDLE_DIRS = ['references', 'scripts', 'assets', 'adapters'];
```
Comment (NicolasMassart) [inline tools/skill-schema.mjs]: **suggestion (non-blocking):** Guard `BUNDLE_DIRS` ↔ `tools/install` against silent drift.

`BUNDLE_DIRS` mirrors a hard-coded Bash list in `tools/install` (the loop that copies `references scripts assets adapters`). The header comment notes the manual sync, which is fine — but a 5-line node test that greps the bash list and asserts equality with `BUNDLE_DIRS.join(' ')` would catch silent drift in the same spirit as making the schema the single source.

```javascript
test('tools/install bundle list matches BUNDLE_DIRS', () => {
  const install = readFileSync(path.resolve(REPO, 'tools/install'), 'utf8');
  const m = /for bundle in ([a-z ]+); do/u.exec(install);
  assert.ok(m, 'could not find bundle loop in tools/install');
  assert.equal(m[1].trim(), BUNDLE_DIRS.join(' '));
});
```

**Review by NicolasMassart — CHANGES_REQUESTED** [review]
Nice direction — reusing `collectSkills` / `parseFrontmatter` is the right shape, since the gate validates exactly what ships. Heads-up: the new workflow is currently failing on this branch and needs a fix before merge.

File: .github/workflows/lint-skill-entry.yml (hunk 1 above)
Comment (MajorLift) [reply .github/workflows/lint-skill-entry.yml]: Done in 105ecf0. Replaced `tj-actions/changed-files` with `git diff -z --name-only`, written to a file, read back with `mapfile -d ''` into an array, and passed as argv. Filenames never reach a command line, so a space, quote, or metacharacter is just a filename. Needed `fetch-depth: 0` for the base commit.

Added a test for it — `a filename containing spaces survives argv handling` in `test/lint-skill-entry.test.mjs`.

File: .github/scripts/lint-skill-entry.mjs (hunk 2 above)
Comment (MajorLift) [reply .github/scripts/lint-skill-entry.mjs]: Done in 105ecf0, taking the sketch as written — `validatePathShape` runs over the changed paths *before* the `collectSkills` filter.

Confirmed the diagnosis rather than assuming it. Running the pre-fix linter against a skill-shaped directory with no `skill.md`:

```
0 skill(s) checked, 0 error(s), 0 warning(s).
exit: 0
```

and the same input after:

```
error:   skill root "domains/testing/skills/ghost/" has no readable skill.md
0 skill(s) checked, 1 error(s), 0 warning(s).
```

All four of your examples are covered.

File: test/lint-skill-entry.test.mjs (hunk 3 above)
Comment (MajorLift) [reply test/lint-skill-entry.test.mjs]: Done in 105ecf0. `lint()` now forwards paths so the tests exercise the branch CI runs, and all seven cases you listed are covered — valid changed `skill.md`, a changed `references/` file mapping back to its root, a malformed `domains/**` path, a missing `skill.md` reported rather than skipped, a filename with spaces, warnings-only exiting 0, and `DESCRIPTION_MAX` exactly versus `+1`.

19 tests in that file now.

File: .github/workflows/lint-skill-entry.yml (hunk 4 above)
Comment (MajorLift) [reply .github/workflows/lint-skill-entry.yml]: Replaced rather than pinned — `git diff -z` in 105ecf0, so `tj-actions/changed-files` is gone.

Code scanning then flagged the same class on the two I had left: `actions/checkout` and `actions/setup-node`, plus credential persistence on checkout. Same answer, 93c4ab6 — this workflow now uses `MetaMask/action-checkout-and-setup@v3` like every other one here, which takes `fetch-depth` and was all raw checkout was providing. No non-MetaMask actions remain in the file.

File: .github/workflows/lint-skill-entry.yml (hunk 5 above)
Comment (MajorLift) [reply .github/workflows/lint-skill-entry.yml:7]: Done in b197ebc, taking the branch option. A change under `tools/` or `.github/scripts/` sets `full_audit=true` and runs the linter with no arguments, so a tightened rule is applied to the whole catalogue rather than to nothing.

That is what surfaced the `workflows/` directories: 14 files across two `web3-tools` skills, referenced 22 times from their bodies, never in the installer bundle list. Fixed in ce975b2 by adding `workflows` to `copy_bundle_dirs` — 22/22 references resolve after install now, verified against a real `web3-tools` install.

File: tools/skill-schema.mjs (hunk 6 above)
Comment (MajorLift) [reply tools/skill-schema.mjs]: Done in b197ebc, and it was in three files rather than one — `README.md`, `CONTRIBUTING.md`, and `.github/SKILL_TEMPLATE.md` all said 1,536. All now say 1,024, with the reason recorded once in the README: it is the lowest ceiling across operators, so a description that fits is accepted by all of them.

File: .github/scripts/lint-skill-entry.mjs (hunk 7 above)
Comment (MajorLift) [reply .github/scripts/lint-skill-entry.mjs:71]: Done in b197ebc. `SCOPE_VALUES` added to the schema; `scope` outside it and `mandatory` that is neither truthy nor falsy both warn, naming the fallback the author will actually get (`installs as project scope`, `treated as false`). Warnings rather than errors, as you suggested, so the blocking surface stays where it is.

File: .github/scripts/lint-skill-entry.mjs (hunk 8 above)
Comment (MajorLift) [reply .github/scripts/lint-skill-entry.mjs]: Done in b197ebc — `collectSkills([ROOT], '*')` with a comment noting the linter never reads `repoApplicable`, so the reliance is deliberate rather than incidental.

File: .github/scripts/lint-skill-entry.mjs (hunk 9 above)
Comment (MajorLift) [reply .github/scripts/lint-skill-entry.mjs:122]: Fixed in b197ebc, but not by anchoring — that produced false positives, so I want to flag the deviation rather than bury it.

Your diagnosis is right: the trailing `\b` let `## When To Use Cases` satisfy `When To Use`. Anchoring to end-of-line fixes that and rejects four headings that are genuinely the section:

```
2 × ## Workflows
2 × ## Workflow (interactive)
```

Meanwhile `## When To Use Cases` does not appear anywhere in the corpus, so anchoring traded a hypothetical false negative for four real false positives.

The pattern now allows an optional plural and an optional parenthetical qualifier, nothing else:

```
^#{1,4}\s+${section}s?(?:\s*\([^)]*\))?\s*$
```

Your case is still rejected — verified against a fixture with `## When To Use Cases`, which warns — and warnings are back to 84 rather than 88. Happy to switch to a strict anchor and rename the four headings instead if you would rather the rule were exact.

File: package.json (hunk 10 above)
Comment (MajorLift) [reply package.json]: Renamed to `audit:skills` in b197ebc, with `CONTRIBUTING.md` and `README.md` updated to match. With no arguments it audits the whole catalogue rather than linting a change, so the name now says so.

The post-merge failure you anticipated no longer applies: the three errors the audit surfaced are fixed in ce975b2 (`workflows/` bundled, `metadata` accepted, `performance` description trimmed 1078 → 928), so a full audit on `main` reports 0 errors across 46 skills. The 84 remaining warnings are all missing recommended sections, which do not fail.

File: test/lint-skill-entry.test.mjs (hunk 11 above)
Comment (MajorLift) [reply test/lint-skill-entry.test.mjs:106]: Partly overtaken, so I went one level up. The `workflows/` drift is fixed rather than gated — `copy_bundle_dirs` now ships it (ce975b2), because those 14 files were referenced 22 times and never delivered. Testing that `workflows/` fails would now assert the wrong thing.

So the test asserts the general guarantee instead (b197ebc): a sibling directory the installer does not copy fails (`playbooks/`), and every directory in `BUNDLE_DIRS` is accepted. The second is the one that would have caught the original bug, since it fails the moment the schema and the installer disagree.

File: tools/skill-schema.mjs (hunk 12 above)
Comment (MajorLift) [reply tools/skill-schema.mjs]: Done in ce975b2, close to your sketch: `BUNDLE_DIRS matches the directories tools/install copies` greps the bash loop and asserts set equality.

Added a second in the same spirit — `the schema declares every frontmatter key tools/install reads` parses the `frontmatter_value` calls out of the installer and fails on any key the schema does not declare. That direction bit already: the installer reads `mandatory` and `scope`, and an earlier draft of the allowlist omitted both.

**Review by NicolasMassart — APPROVED** [review]
Looks good to me.
